Doc Reorg: Unpackaged Metadata & Metadata ETL (#2069)
SFDO-Tooling/CumulusCI: docs/unpackaged.rst

The task ``deploy_post``, which is part of the ``config_dev``, ``config_qa``, and ``config_managed`` flows, is responsible for deploying these bundles.

Metadata that's not intended to be delivered to all installations of the product should *not* be included in ``unpackaged/post``. It's critical for managed package projects that this metadata include namespace tokens (see :ref:`namespace-injection`).

``unpackaged/config``: Tailoring Orgs
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Projects may come with more than one supported configuration in their CumulusCI automation. For example, projects may support distinct, tailored ``dev_org``, ``qa_org``, and ``install_prod`` flows, each of which performs different setup for their specific use case. Unpackaged metadata stored in ``unpackaged/config`` is a tool to support operational needs to tailor orgs to different configurations. 

> For example, a testing-oriented scratch org may wish to deploy a customized set of Page Layouts to help testers easily visualize data under test. Such page layouts could be stored in ``unpackaged/config/qa``.

Reviewer: Oh, I see that's what is in the "Customize Config Flows" section below.

Unpackaged Metadata Folder Structure
------------------------------------

All unpackaged metadata is stored in the ``unpackaged`` directory tree. Within this directory are the three top-level directories, ``unpackaged/pre``, ``unpackaged/post``, and ``unpackaged/config``.

> These trees contain metadata bundles in Metadata API format: that is, a directory containing a ``package.xml`` manifest and Metadata API-format source code. CumulusCI does not support Salesforce DX format for unpackaged bundles.

Reviewer: Actually since the merge of the 2gp branch, I believe if you point `deploy` at a folder that doesn't have a package.xml, it will use sfdx to convert it in a tempdir to deploy. But this may not work well with namespace tokens in some locations if they get in the way of the conversion. There is potential to make it work with a scheme like this:
1. Copy sfdx format to a tempdir.
2. Do namespace injection.
3. Convert to mdapi.

.. _namespace-injection:

Namespace Injection
-------------------

Projects that build managed packages often must construct their unpackaged metadata to be deployable in multiple contexts: unmanaged deployments, such as developer orgs, unmanaged namespaced scratch orgs, and also managed contexts, such as a beta test org or a demo org created with ``install_prod``.

>   Projects that are building an org implementation (not a package with a namespace) do not need to use namespace injection.

Reviewer: You haven't defined what namespace injection is yet...

Manage Unpackaged Configuration
===============================

Not everything that's part of an application can be part of a package.

CumulusCI implements the Product Delivery Model by offering support for complex applications -- applications that may include multiple managed packages as well as unpackaged metadata, and setup automation that configures org settings or makes precise changes to existing configuration.

The tools used to implement that support are *unpackaged metadata* and *Metadata ETL*. 

Unpackaged metadata refers to metadata that is not delivered as part of a package, and can include both support metadata delivered to users as well as metadata that operationally configures orgs used by the product. 

> Metadata ETL is a suite of tasks that support surgically altering existing metadata in an org. It's a powerful technique that changes the unpackaged configuration in an org without risking damage by overwriting existing customizations with stored metadata. Metadata ETL is relevant for delivering applications to customers safely, and is often a superior alternative to unpackaged metadata.

Reviewer 2: Would we call it a feature rather than a technique? And the end of this sentence is a bit confusing: Do the existing customizations have stored metadata? Or does Metadata ETL use stored metadata to not damage existing customizations? (I’m  95% certain it’s the former, but I would like confirmation.)

> Metadata ETL is a suite of tasks that support surgically altering existing metadata in an org. It's a powerful technique that changes the unpackaged configuration in an org without risking damage by overwriting existing customizations with stored metadata. Metadata ETL is relevant for delivering applications to customers safely, and is often a superior alternative to unpackaged metadata.

Reviewer 2: Would we call it a feature rather than a technique? And the end of this sentence is a bit confusing: Do the existing customizations have stored metadata? Or does Metadata ETL use stored metadata to not damage existing customizations? (I’m 95% certain it’s the former, but I would like confirmation.)

Reviewer 3: [ETL](https://en.wikipedia.org/wiki/Extract,_transform,_load) is an established technique/procedure/methodology. I wouldn't refer to it as a feature.

Ya, this sentence could be better. Maybe: "It's a powerful technique that alters the unpackaged configuration in an org without risking damage to existing customization by overwriting them with incoming metadata."

Let me know if that's any better.

Author: Yes to the Brandon edit!

----------------------------

``unpackaged/pre``: Prepare an Org
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some projects require that unpackaged metadata be deployed to finish the customization of an org *before* the project's own code and metadata are deployed.

    Example: The Nonprofit Success Pack (NPSP) must deploy unpackaged Record Types prior to installing its own packages. ``unpackaged/pre`` is the location designed for such metadata, which is stored in subdirectories such as ``unpackaged/pre/first``.

CumulusCI's standard flows that build orgs, such as ``dev_org`` and ``install_prod``, always deploy metadata bundles found in ``unpackaged/pre`` before proceeding to the deployment of the application. It's also easy to include ``unpackaged/pre`` metadata in customer-facing installers run via MetaDeploy.

The ``deploy_pre`` task, which is part of the ``dependencies`` flow, is responsible for deploying these bundles.

.. important:: Do not include metadata in ``unpackaged/pre`` unless it is intended to be delivered to *all* installations of the product.

``unpackaged/post``: Configuration After Package Install
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Projects often include metadata that is genuinely part of the application, but either cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the project's own code and metadata are deployed first and the org is configured.

Reviewer 2: Changed up some of the wording in the first sentence and added the second one. Please confirm that everything is correct.

Reviewer 3: ```suggestion
Projects often include metadata that is genuinely part of the application, but either cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the package's own code and metadata are deployed first and the org is configured.
```

> Projects often include metadata that is genuinely part of the application, but either cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the project's own code and metadata are deployed first and the org is configured.

Author: ```suggestion
Projects often include metadata that is genuinely part of the application, but cannot be delivered as part of a managed package for operational reasons. This metadata must be deployed *after* the project's own code and metadata are deployed first and the org is configured.
```

Unpackaged metadata stored in ``unpackaged/config`` is a tool to support operational needs that tailor orgs to different configurations. 

    Example: A testing-oriented scratch org needs to deploy a customized set of Page Layouts to help testers easily visualize data under test. Such page layouts are stored in ``unpackaged/config/qa``.

Unpackaged Metadata Folder Structure
------------------------------------

All unpackaged metadata is stored in the ``unpackaged`` directory tree, which contains these top-level directories.

* ``unpackaged/pre``
* ``unpackaged/post``
* ``unpackaged/config``

These trees contain metadata bundles in Metadata API format, represented as a directory containing a ``package.xml`` manifest and Metadata API-format source code. CumulusCI does not support Salesforce DX format for unpackaged bundles.

Namespace Injection
-------------------

Projects that build managed packages often construct their unpackaged metadata to be deployable in multiple contexts.

* Unmanaged deployments, such as developer orgs
* Unmanaged namespaced scratch orgs
* Managed contexts, such as a beta test org or a demo org created with ``install_prod``

> Because projects that are building an org implementation or a non-namespaced package do not have a namespace, or a distinction between managed and unmanaged contexts, these projects must also build namespaced packages.

Reviewer 2: Why must they build namespaced packages? I would like to cap this sentence with just a few words.

"These projects must also build namespaced packages to __________________."

Reviewer 3: @davidmreed - Is a non-namespaced package just another way of saying 'unamangeg/unlocked' package? Also wondering about Justin's question here. The following example outlines how `unpackaged/post` needs to work in both managed and unmanaged contexts; however, if I'm building an org implementation in a package without a namespace, what's the reason for needing to build namespaced packages?

Author: I think I wrote this sentence, but it's total nonsense. 

```suggestion
Because projects that are building an org implementation or a non-namespaced package do not have a namespace, or a distinction between managed and unmanaged contexts, these projects typically don't need to use namespace injection.
```

>     Example: Metadata located in ``unpackaged/post`` is deployed after the application code in both unmanaged and managed contexts. If that metadata contains references to the application components, it must be deployable when that metadata is nampespaced (in a managed context or namespaced scratch org) *and* when it is not (in an unmanaged context).

Reviewer 2: This is getting hard to follow simply because I don’t fully know why things are required to be namespaced. Which is okay because I'm still quite the novice when it comes to building anything with ``cci``. That said, is it safe to presume that most devs this deep into the material know about namespaces, why they're needed and will follow this material with (relative) ease?

Reviewer 3: This example is clear to me (not sure how helpful this is), but I think it should be safe to assume that people reading this are going to be interested in, or familiar with, Salesforce packages in general. We might want to change this example based on feedback from David R. regarding your prior question.

Author: I'm happy to expand on or clarify this example but I think it's an important one because it is very common.

Reviewer 3: Let's keep this then.

>     Example: Metadata located in ``unpackaged/post`` is deployed after the application code in both unmanaged and managed contexts. If that metadata contains references to the application components, it must be deployable when that metadata is nampespaced (in a managed context or namespaced scratch org) *and* when it is not (in an unmanaged context).

Author: ```suggestion
    Example: Metadata located in ``unpackaged/post`` is deployed after the application code in both unmanaged and managed contexts. If that metadata contains references to the application components, it must be deployable when that metadata is namespaced (in a managed context or namespaced scratch org) *and* when it is not (in an unmanaged context).
```

> CumulusCI uses a strategy called *namespace injection* to support this use case. Namespace injection is very powerful, and requires care from application implementors to ensure that metadata remains deployable in all contexts.

Reviewer 2: "application implementors"

Just confirming this is what we want to call them because this is the first time I've heard this label being used.

Reviewer 3: We could maybe use "application developers and architects" as a substitute?

Author: ```suggestion
CumulusCI uses a strategy called *namespace injection* to support this use case. Namespace injection is very powerful, and requires care from the project team to ensure that metadata remains deployable in all contexts.
```

* ``%%%NAMESPACED_ORG%%%`` is replaced with the package’s namespace in a namespaced org *only*, not in a managed installation. Otherwise, it remains blank.
    .. note:: This token supports use cases where components in one unpackaged metadata bundle refer to components in another, and the dependency bundle acquires a namespace by being deployed into a namespaced org.
* ``%%%NAMESPACE_OR_C%%%`` is replaced with the package’s namespace in any context with a namespace (such as a namespaced org or managed org). Otherwise, it is replaced with ``c``, the generic namespace used in Lightning components.
* ``%%%NAMESPACED_ORG_OR_C%%%`` is replaced with the package's namespace in a namespaced org *only*, not in a managed installation. Otherwise, it is replaced with ``c``, the generic namespace used in Lightning components.
* ``%%%NAMESPACE_DOT%%%`` is replaced with the package’s namespace in any context with a namespace (such as a namespaced org or managed org) followed by a period (``.``) rather than two underscores.
    .. note:: This token is used to construct references to packaged Record Types.

..

>     Example: A portion of metadata from the Nonprofit Success Pack (NPSP) is stored in a subdirectory under ``unpackaged/post``, meaning it's deployed after the application metadata. This metadata updates a Compact Layout on the ``Account`` object, and references packaged metadata from the application as well as from other managed packages. To complete these tasks, this metadata requires the use of namespace tokens to represent the ``npsp`` namespace, letting CumulusCI automatically adapt the metadata to deploy into managed and unmanaged contexts.

Reviewer 2: “To complete these tasks...”

Added this for clarity, so I’m simply confirming that the updating and referencing are indeed tasks. And if not, is there a better word? Steps?

Reviewer 3: Hm. Not sure that we need that there. The "update" and "referencing" in the paragraph are referring to characteristics of particular piece of metadata, not things that will be done in the future. When deployed, the metadata updates a Compact Layout on the Account object, and it also has references to another managed package (in the form of the `npo02__TotalOppAmount__c` field).

Author: "Steps" would be fine. "Tasks" is definitely confusing and ambiguous.

.. code-block:: xml

    <?xml version="1.0" encoding="UTF-8"?>
    <CustomObject xmlns="http://soap.sforce.com/2006/04/metadata">
        <compactLayouts>
            <fullName>NPSP_Household_Account</fullName>
            <fields>Name</fields>
            <fields>npo02__TotalOppAmount__c</fields>
            <fields>%%%NAMESPACE%%%Number_of_Household_Members__c</fields>
            <label>NPSP Household Account</label>
        </compactLayouts>
    </CustomObject>

..

    Note that only the reference to the NPSP field ``Number_of_Household_Members__c`` is tokenized. (This field is called ``npsp__Number_of_Household_Members__c`` when installed as part of the managed package.) References to NPSP's own managed package dependency, ``npo02``, are not tokenized because this metadata is always namespaced when installed.

> 

Reviewer 2: Why this specific field? Also, is this detail in parentheses relevant/helpful for a dev to know for this example to make sense? I honestly can't make heads or tails of it.  To be honest, it feels like we’re currently missing one or two details here about why only this field is the one to get tokenized.

Author: There's no specific importance to this field, it's just a good example.

But if it's not clear why this field is tokenized, the explanation so far has failed. Should we talk through it? n.b. this are is _notoriously_ difficult to understand for users and even release engineers.

Reviewer 3: Salesforce has a convention of using `namespace__` in front of the field name to indicate which package a particular field belongs to. This particular field is the only one in the compact layout that is a member of the `npsp` package (this is why it's the only one getting the token). We know this only because we were told so here. 

The detail in the parens does help I believe, it's letting us know that in a "managed" context this field will be display as `nspsp__Number_of_Household_Members__c` in an org. This is useful because then I can see how the `%%%NAMESPACE%%%` token is getting replaced with `npsp__` in a "managed" context.

    If this metadata isn't tokenized, it fails to deploy into an org containing NPSP as a beta or released managed package.

> 

Reviewer 2: Why? This example might need more clarification, or it might need to be swapped out with one easier to comprehend.

(Or it might be completely fine for an advanced dev. Your call.)

Author: ```suggestion
    If this metadata isn't tokenized, it fails to deploy into an org containing NPSP as a beta or released managed package — because in that context, the field `Number_of_Household_Members__c` is namespaced, as `npsp__ Number_of_Household_Members__c`, and we must refer to it as such.

```

Reviewer 3: We are pretty deep in the weeds here. We should assume that users reading this section are likely a release engineer familiar with packaging basics. But I think we could maybe make this a little more explicit in how the error actually occurs.

@davidmreed - I need you help on how to finish the example. I believe I'm correct up until "the field would be deployed" part. Is the error on deploy, or is this more related with a problem that occurs in the `package.xml` file, or something else?

```suggestion
    If this metadata isn't tokenized, it will fail to deploy into an org containing NPSP as a beta or released managed package. Since all metadata is retrieved in an unmanaged context (i.e. without a namespace prefix) during development, if there is no `%%%NAMESPACE%%%` token present, the field would be deployed without a namespace, and thus result in an error.
```

> .. note:: The resolution of component references in namespaced scratch orgs and in managed installations of the same metadata are not identical. Metadata that is tokenized and deploys cleanly in a namespaced scratch org can still fail in a managed context.

Reviewer 2: This seems like real inside baseball. Is it reasonable to presume that advanced devs will understand this? (Again, it's completely fine if I don't because I presume at this point that only advanced devs would be this far into the written materials.)

Author: This is Deep Magic. I think it's an important note but it's okay that not everyone gets it. Those who are operating with these tools (namespaced scratch orgs + managed packages) will hopefully get it.

Reviewer 3: I think this should be fine. There's a lot of jargon, but a release engineer or advanced dev is going to be the target role for this content.

    Example: If you have changes to make in the application as well as in unpackaged metadata, complete the application changes first, retrieve them, and then make the unpackaged changes and retrieve those. If you conflate changes to components that live in separate elements of your project, it's difficult to untangle them.
* Whenever possible, build your unpackaged metadata in an org that contains a beta or released managed package. By doing so, the metadata contains namespaces when extracted, which CumulusCI easily replaces with tokens when retrieving metadata. It's difficult to manually tokenize metadata that's retrieved from an unmanaged org without namespaces. 

After building changes to unpackaged metadata in a managed org, retrieve it using ``retrieve_changes`` with the additional ``namespace_tokenize`` option, and use the ``path`` option to direct the retrieved metadata to your desired unpackaged directory.

    Example: Run ``retrieve_changes`` to retrieve metadata changes into the ``unpackaged/config/qa`` subdirectory, and replace references to the namespace ``npsp`` with the appropriate token.

.. code-block:: console

>     $ cci task run retrieve_changes --path unpackaged/config/qa --namespace_tokenize npsp

Reviewer 2: Swapped out the old -o flags. Please confirm this updated command is correct.

Author: This looks correct to me.

Reviewer 3: 10-4
